File: src/common/net/auoeke/wheel/loader/TransformingClassLoader.java
package net.auoeke.wheel.loader;

import java.net.URL;
import java.security.CodeSource;
import java.security.cert.Certificate;
import java.util.ArrayList;
import java.util.Collection;
import java.util.IdentityHashMap;
import java.util.List;
import java.util.Map;
import org.gradle.internal.classloader.VisitableURLClassLoader;
import org.objectweb.asm.ClassReader;
import org.objectweb.asm.ClassWriter;
import org.objectweb.asm.tree.ClassNode;
import user11681.reflect.Classes;
import user11681.uncheck.Uncheck;

public class TransformingClassLoader extends VisitableURLClassLoader {
    public static final long klass = Classes.klass(TransformingClassLoader.class);

    private static final List<Map.Entry<String, Transformer>> transformers = new ArrayList<>();
    private static final Map<TransformingClassLoader, DelegateTransformingLoader> delegates = new IdentityHashMap<>();

    public TransformingClassLoader(String name, ClassLoader parent, Collection<URL> urls) {
        super(name, parent, urls);
    }

    public void transform(String name, Transformer transformer) {
        Class<?> foundClass = this.findLoadedClass(name);

        if (foundClass == null) {
            transformers.add(Map.entry(name, transformer));
        } else {
            delegates.computeIfAbsent(this, DelegateTransformingLoader::new).define(name, false, transformer);
        }
    }

    @Override
    public Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
        if (!name.startsWith("java.")) {
            DelegateTransformingLoader delegate = delegates.get(this);

            if (delegate != null) {
                Class<?> override = delegate.findLoadedClass(name);

                if (override != null) {
                    return override;
                }
            }

            for (int index = transformers.size() - 1; index >= 0; index--) {
                Map.Entry<String, Transformer> entry = transformers.get(index);

                if (name.equals(entry.getKey())) {
                    transformers.remove(index);

                    return this.define(name, resolve, entry.getValue());
                }
            }
        }

        return super.loadClass(name, resolve);
    }

    @SuppressWarnings("ConstantConditions")
    protected Class<?> define(String name, boolean resolve, Transformer transformer) {
        URL classFile = this.getResource(name.replace('.', '/') + ".class");
        ClassNode node = new ClassNode();
        Uncheck.handle(() -> new ClassReader(classFile.openStream()).accept(node, 0));
        transformer.transform(node);

        ClassWriter writer = new ClassWriter(ClassWriter.COMPUTE_FRAMES);
        node.accept(writer);
        byte[] bytecode = writer.toByteArray();
        Class<?> type = this.defineClass(name, bytecode, 0, bytecode.length, new CodeSource(classFile, (Certificate[]) null));

        if (resolve) {
            this.resolveClass(type);
        }

        return type;
    }
}

File: src/net/auoeke/wheel/loader/TransformingClassLoader.kt
package net.auoeke.wheel.loader

import net.auoeke.extensions.type
import net.auoeke.reflect.Classes
import org.gradle.internal.classloader.VisitableURLClassLoader
import org.objectweb.asm.ClassReader
import org.objectweb.asm.ClassWriter
import org.objectweb.asm.tree.ClassNode
import java.net.URL
import java.security.CodeSource
import java.security.cert.Certificate
import java.util.*

@Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS", "RECEIVER_NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
open class TransformingClassLoader(name: String, parent: ClassLoader?, urls: Collection<URL?>?) : VisitableURLClassLoader(name, parent, urls) {
    fun transform(name: String, transformer: (ClassNode) -> Unit) {
        val foundClass = this.findLoadedClass(name)

        if (foundClass === null) {
            transformers.add(name to transformer)
        } else {
            delegates.computeIfAbsent(this, ::DelegateTransformingLoader).define(name, false, transformer)
        }
    }

    public override fun loadClass(name: String, resolve: Boolean): Class<*> {
        if (name.startsWith("java.") || name.startsWith("kotlin.")) {
            return super.loadClass(name, resolve)
        }

        delegates[this]?.findLoadedClass(name)?.also {return it}

        for (index in transformers.indices.reversed()) {
            val (key, value) = transformers[index]

            if (name == key) {
                transformers.removeAt(index)

                return this.define(name, resolve, value)
            }
        }

        return super.loadClass(name, resolve)
    }

    private fun define(name: String, resolve: Boolean, transformer: (ClassNode) -> Unit): Class<*> {
        val classFile = this.getResource("${name.replace('.', '/')}.class")
        val node = ClassNode()
        ClassReader(classFile.openStream()).accept(node, 0)
        transformer(node)

        val writer = ClassWriter(ClassWriter.COMPUTE_FRAMES)
        node.accept(writer)
        val bytecode = writer.toByteArray()
        val type = this.defineClass(name, bytecode, 0, bytecode.size, CodeSource(classFile, null as Array<Certificate?>?))

        if (resolve) {
            this.resolveClass(type)
        }

        return type
    }

    companion object {
        val klass = Classes.klass(type<TransformingClassLoader>())
        private val transformers: MutableList<Pair<String, (ClassNode) -> Unit>> = ArrayList()
        private val delegates: MutableMap<TransformingClassLoader, TransformingClassLoader> = IdentityHashMap()

        init {
            Classes.load<Any>("kotlin.jvm.internal.Intrinsics")
            Classes.load<Any>("kotlin.text.StringsKt")
        }
    }
}
